### PR TITLE
test(firestore): error handling for loadBundle() & namedQueryGet()

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

Requires this test [API PR to be merged](https://github.com/invertase/react-native-firebase-tests-api/pull/4) first.

Follow up PR to test error handling for `namedQueryGet()` & `loadBundle()` API. 

## Related Issues

none.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:154]: No, I tried that but it think it might be deprecated.

Comment (mikehardy) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:154]: I was curious about how to work around language mismatches just as a language thing, so I looked - I'm curious if this works in practice as well :nerd_face:

Comment (Salakar) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:154]: Since it is a Promise itself we should just use the [second arg of `then` (`onReject`)](https://firebase.google.com/docs/reference/js/firebase.firestore.LoadBundleTask#then), e.g.:

```dart
  external PromiseJsImpl then([
    void Function(dynamic) onResolve,
    void Function(dynamic) onReject, // <----
  ]);
```

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: `Query` probably should still be non-nullable

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:161]: Don't we need to handle "value" here?

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:166]: Some trailing commas:
```suggestion
            stackTrace: StackTrace.fromString(error.stack),),);
```

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore_interop.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore_interop.dart:261]:
```suggestion
  external PromiseJsImpl then([
    Func1? onResolve,
    Function(FirestoreError) onReject,
  ]);
```

File: packages/cloud_firestore/cloud_firestore_web/lib/src/load_bundle_task_web.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/load_bundle_task_web.dart:8]: This isn't a broadcast stream anymore. I don't think that's voluntary

File: packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart:79]: Could you use expectLater/throwsA instead? :pray:

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: even if it can be `null`? There is also a `null` check [just below it](https://github.com/FirebaseExtended/flutterfire/blob/00f0ab6f0c6b29c96e9664e0d5939ba5c7bb033f/packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart#L118-L124).

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:161]: Already handled in the `onProgress` [event handler](https://github.com/FirebaseExtended/flutterfire/blob/00f0ab6f0c6b29c96e9664e0d5939ba5c7bb033f/packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart#L156).

File: packages/cloud_firestore/cloud_firestore_web/lib/src/load_bundle_task_web.dart
Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/load_bundle_task_web.dart:8]: I don't understand what is required? 😅

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (rrousselGit) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:161]: Are you sure that `.then` won't emit something _after_ `onProgress`

Like onProgress sending the 99% completion then .then sending the 100%

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (rrousselGit) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: But the function cannot return null. The null-check throws an exception

File: packages/cloud_firestore/cloud_firestore_web/lib/src/load_bundle_task_web.dart
Comment (rrousselGit) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/load_bundle_task_web.dart:8]: It's missing an `asBroadcastStream` (with a pause/resume we discussed before too)

And possibly a test for that

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: this test [indicates](https://github.com/FirebaseExtended/flutterfire/blob/50183df79f73c0840b9330022064c78f462db012/packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart#L122) that it does return `null` 😄.

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:161]: positive.

Comment (rrousselGit) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:161]: In that case, lgtm

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (rrousselGit) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: I'm wondering if we're talking about the same thing.

It's about `Future<Query?>` into `Future<Query>`

Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: Yeah, talking about the same thing. I assume I'm misunderstanding, but if the result of the future can be `null`. Then the type has to be `Future<Query?>`, right? The test also proves it as the `query == null` condition throws the error necessary for the test to pass.

Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: Just to be clear, I have also tested on `namedQuery` on web. It will return `null` if the named query does not exist.

Comment (russellwheatley) [reply packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:114]: ok, ignore me, I'm being a dummy.

File: packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart
Comment (Salakar) [inline packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart:88]:
```suggestion
        // will listen & pause after first event received
```

File: packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart:79]: Could you test the error code?

File: packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart
Comment (Salakar) [inline packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart:96]:
```suggestion
        // will resume & pause after second event received
```

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (Salakar) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:155]:
```suggestion
          // error handled in addError() call below.
```

File: packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart:92]:
```suggestion
        await expectLater(
          task.stream,
          emits(isA<LoadBundleTaskSnapshot>().having((ts) => ts.taskState,
            'taskState', LoadBundleTaskState.running)),
        );
```

**Review by Salakar — APPROVED** [review]
LGTM pending final approval from Remi, some minor suggestions on code comment spacing

File: packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore/example/test_driver/load_bundle_e2e.dart:103]:
```suggestion
                    LoadBundleTaskState.success)),),);
```

File: packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore.dart:123]: This needs an error code, so that people can differentiate this exception from others

**Comment by rrousselGit** [comment]
Added some comments about the error handling, otherwise LGTM

**Review by rrousselGit — APPROVED** [review]

**Pull request merged**
